File: js/src/c3OtimesC3/upb/tilesUpb.ts
/**
 * https://arxiv.org/abs/quant-ph/9808030
 *
 * C. H. Bennett, D. P. DiVincenzo, T. Mor, P. W. Shor, J. A. Smolin, B. M.
 * Terhal, "Unextendible Product Bases and Bound Entanglement", Phys. Rev.
 * Lett. 82, 5385 (1999).
 */

import type { Matrix } from 'mathjs';
import { tile, upb } from '../../utils/index.js';

/**
 * The five product vectors of the "Tiles" UPB on C³ ⊗ C³.
 *
 * @returns the five normalized product vectors forming the UPB.
 */
export function tilesBasis(): Matrix[] {
  return [0, 1, 2, 3, 4].map(tile);
}

/**
 * The bound entangled state built from the Tiles UPB.
 *
 * @returns the bound entangled state on the orthogonal complement of the UPB.
 */
export function tilesUpb(): Matrix {
  return upb(tilesBasis());
}

File: python/src/bound_entangled/c3_otimes_c3/upb/tiles_UPB.py
"""https://arxiv.org/abs/quant-ph/9808030

C. H. Bennett, D. P. DiVincenzo, T. Mor, P. W. Shor, J. A. Smolin, B. M.
Terhal, "Unextendible Product Bases and Bound Entanglement", Phys. Rev. Lett.
82, 5385 (1999).
"""

import numpy as np
from toqito.states import tile
from bound_entangled.utils import upb


def tiles_upb() -> np.ndarray:
    """The bound entangled state built from the "Tiles" UPB on C^3 (x) C^3.

    Returns:
        np.ndarray: the bound entangled state on the orthogonal complement of
        the Tiles unextendible product basis (`toqito.states.tile`).

    Examples:
        >>> from toqito.matrix_props import is_density
        >>> from toqito.state_props import is_ppt, is_separable
        >>> state = tiles_upb()
        >>> is_density(state)
        True
        >>> is_ppt(state, dim=[3, 3])
        True
        >>> sep, _ = is_separable(state, dim=[3, 3])
        >>> sep
        False
    """
    return upb([tile(i) for i in range(5)])
